pkg/client/client_v2.go

```diff
@@ -510,6 +520,24 @@ func (c *clientV2) removeUnexpectedUnits(expected *proto.CheckinExpected) int {
 	return removedCount
 }
 
+// for now the component configuration is applied to the whole process globally
+// when there is a need of propagating changes we should implement an update channel
+// similar to `UnitChanges() <-chan UnitChanged`, however, this would be a breaking change
+// because a user of the client will have to read from this channel,
+// otherwise the client would stay blocked.
+func (c *clientV2) syncComponent(expected *proto.CheckinExpected) {
+	if expected.Component == nil {
+		return
+	}
+	c.componentMu.Lock()
+	defer c.componentMu.Unlock()
+
+	if expected.Component.Limits.MaxProcs > 0 {
+		_ = runtime.GOMAXPROCS(int(expected.Component.Limits.MaxProcs))
+	}
```

Comment:
If MaxProcs is set to a value < runtime.NumCPU and then the setting is removed the policy, this will just leave the setting as the previous value.

I think it would be much more convenient if removing the setting from the policy or setting it to zero returned it to the default value of runtime.NumCPU. 

Otherwise a user has no way to go back to the default setting because there is no guarantee that all the agents in a policy are executing on machines with the same number of CPUs. There is no one value they can set to go back to the number of CPUs on the machine.


This call also may not be free from a performance perspective, it locks the global scheduler lock on each call. The implementation is in https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/runtime/debug.go;l=16

```go
func GOMAXPROCS(n int) int {
	if GOARCH == "wasm" && n > 1 {
		n = 1 // WebAssembly has no threads yet, so only one CPU is possible.
	}

	lock(&sched.lock)
	ret := int(gomaxprocs)
	unlock(&sched.lock)
	if n <= 0 || n == ret {
		return ret
	}

	stopTheWorldGC(stwGOMAXPROCS)

	// newprocs will be processed by startTheWorld
	newprocs = int32(n)

	startTheWorldGC()
	return ret
}
```

It may be best to only call this function when we have to. It should be cheap for us to save the most recently applied value and skip this call if it isn't needed.